xonsh/history.py

```diff
@@ -229,6 +232,177 @@ def i_am_at_the_front(self):
+    return [(c, t, ind) for ind, (c, t) in commands]
+
+
+def _zsh_hist_formatter(location=None):
+    if not location:
+        location = os.path.join('~', '.zsh_history')
+    z_hist_formatted = list()
+    z_path = os.path.expanduser(location)
+    if os.path.isfile(z_path):
+        with open(z_path, 'r') as z_file:
+            z_txt = z_file.read()
+            z_txt = z_txt.encode('utf-8', 'replace').decode('utf-8')
+            z_hist = z_txt.splitlines()
+            try:
+                if z_hist:
+                    for ind, line in enumerate(z_hist):
+                        start_time, command = line.split(';')
+                        try:
+                            start_time = float(start_time.split(':')[1])
+                        except ValueError:
+                            start_time = -1
+                        z_hist_formatted.append((command, start_time, ind))
+                    return z_hist_formatted
+            except Exception as e:
+                print("There was a problem parsing {}.".format(z_path))
+                raise e
+                return None
+
+    else:
+        print("No zsh history file found at: {}".format(z_path))
+        return None
+
+
+def _bash_hist_formatter(location=None):
+    if not location:
+        location = os.path.join('~', '.bash_history')
+    bash_hist_formatted = list()
+    b_path = os.path.expanduser(location)
+    if os.path.isfile(b_path):
+        try:
+            with open(b_path, 'r') as bash_file:
+                b_txt = bash_file.read()
+                b_txt = b_txt.encode('utf-8', 'replace').decode('utf-8')
+                bash_hist = b_txt.splitlines()
+                if bash_hist:
+                    for ind, command in enumerate(bash_hist):
+                        bash_hist_formatted.append((command, 0.0, ind))
+                    return bash_hist_formatted
+        except:
+            print("There was a problem parsing {}.".format(b_path))
+            return None
+    else:
+        print("No bash history file found at: {}".format(b_path))
+        return None
+
+
+@lru_cache()
+def _create_parser():
+    """Create a parser for the "history" command."""
+    p = argparse.ArgumentParser(prog='history',
+                                description='Tools for dealing with history')
+    subp = p.add_subparsers(title='action', dest='action')
+    # show action
+    session = subp.add_parser('session',
+                              help='displays session history, default action')
+    session.add_argument('-r', dest='reverse', default=False,
+                         action='store_true',
+                         help='reverses the direction')
+    session.add_argument('n', nargs='?', default=None,
+                         help='display n\'th history entry if n is a simple '
+                              'int, or range of entries if it is Python '
+                              'slice notation')
+    # 'id' subcommand
```

Comment:
I use _create_parser() in History.show() and so I put it above the class definition for History. Not sure if that was necessary.